### PR TITLE
Implement umode API and requests between hypervisor and umode.

### PR DESCRIPTION
**Pull request by glg-rv**
This PR introduces a umode API library, compiled both on hypervisor and umode, whose goal is to allow data to be passed. There will be two kind of ways in which data will be passed between the two components: registers and memory. This patch implement the register passing part, that happens through `ecall`s.

This API is then used to test sending NOP and HELLO request to umode. Testing initialization and service loop of the umode binary.

### CODE REVIEW

File: src/umode.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+// Entry for umode task.
+static UMODE_ENTRY: Once<u64> = Once::new();
+
+/// Represents a U-mode state with its running context.
+pub struct UmodeTask {
+    arch: RefCell<UmodeCpuArchState>,
+}
+
+impl UmodeTask {
+    /// Initialize U-mode tasks. Must be called once bofore `setup_this_cpu()`.
```
Comment (atulkharerivos) [inline src/umode.rs:250]: Nit: Typo

File: src/umode.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        // Unwrap okay: this is called after `Self::init()`.
+        arch.umode_regs.sepc = *UMODE_ENTRY.get().unwrap();
+        let task = UmodeTask {
+            arch: RefCell::new(arch),
+        };
+        // Install umode in the current cpu.
+        PerCpu::this_cpu().set_umode_task(task);
+    }
+
+    /// Return this CPU's task. Must be call after `Self::setup_this_cpu()`.
```
Comment (atulkharerivos) [inline src/umode.rs]: General comment: Can we express this as a type constraint using enums?

File: umode-api/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#![no_std]
+
+//! Salus Hypervisor <-> U-mode API.
+//!
+//! This library contains data structure that are passed between
```
Comment (atulkharerivos) [inline umode-api/src/lib.rs:9]: Nit: s/structure/structures/

File: src/umode.rs (hunk 2 above)
Comment (glg-rv) [reply src/umode.rs]: 'this' being?

File: umode-api/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+/// Trait to transform a type into A-registers when a set of registers will always transform into
+/// this type.
+pub trait IntoRegisters {
+    /// Get current type from a set of registers.
+    fn from_registers(regs: &[u64]) -> Self;
+    /// Write `self` into a set of registers.
+    fn set_registers(&self, regs: &mut [u64]);
+}
+
+/// Trait to transform a type into A-registers whena set of registers might not be able to be
```
Comment (atulkharerivos) [inline umode-api/src/lib.rs:46]: Nit: Typo

File: src/umode.rs (hunk 2 above)
Comment (atulkharerivos) [reply src/umode.rs]: Ah...sorry, should have been explicit. Basically, was thinking of the lines of the following that's used to represent VmState. The idea is that some methods are available only on certain enum states, and so it's impossible to get the state machine wrong (thereby enforcing the call to Self::setup_this_cpu() for example in the type system).

`pub enum VmStateFinalized {}
/// Represents a finalized, or runnable, VM.
pub type FinalizedVm<'a, T> = VmRef<'a, T, VmStateFinalized>;`

File: src/umode.rs (hunk 1 above)
Comment (glg-rv) [reply src/umode.rs]: Fixed.

File: src/umode.rs (hunk 2 above)
Comment (glg-rv) [reply src/umode.rs]: Oh I see what you meant now. Well, `init` and `setup_this_cpu` are called only once at boot. It is not a dynamic state that has to be tracked.

File: umode-api/src/lib.rs (hunk 4 above)
Comment (glg-rv) [reply umode-api/src/lib.rs]: Heh. Or I was typing with a stereotypical Italian accent. :-) Fixed!

File: Cargo.toml [hunk 5]
```diff
@@ -31,6 +31,7 @@ riscv_page_tables = { path = "./riscv-page-tables" }
 riscv_pages = { path = "./riscv-pages" }
 riscv_regs = { path = "./riscv-regs" }
 s_mode_utils = { path = "./s-mode-utils" }
+umode_api = { path = "./umode-api" }
```
Comment (abrestic-rivos) [inline Cargo.toml:34]: Sorry, the inconsistent hyphenation of `s-mode` vs `umode` kind of bothers me :)

File: umode-api/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+// All types that can be passed in registers must implement `IntoRegisters` or `TryIntoRegisters`.
+
+/// Trait to transform a type into A-registers when a set of registers will always transform into
+/// this type.
+pub trait IntoRegisters {
+    /// Get current type from a set of registers.
+    fn from_registers(regs: &[u64]) -> Self;
+    /// Write `self` into a set of registers.
+    fn set_registers(&self, regs: &mut [u64]);
```
Comment (abrestic-rivos) [inline umode-api/src/lib.rs:43]: nit: I'd call this `to_registers` as that sounds like the more natural complement to `from_registers`.

File: umode-api/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#![no_std]
+
+//! Salus Hypervisor <-> U-mode API.
+//!
+//! This library contains data structures that are passed between
+//! hypervisor and umode.
```
Comment (abrestic-rivos) [inline umode-api/src/lib.rs:10]: Expanding this comment to explain the calling convention would be helpful.

File: umode-api/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+// HypCall: calls from umode to hypervisor.
+
+/// Calls from umode to the hypervisors.
+pub enum HypCall {
+    /// Panic and exit immediately.
+    Panic,
+    /// Print a character for debug.
+    PutChar(u8),
+    /// Return result of previous request and wait for next operation.
+    NextOp(Result<(), Error>),
```
Comment (abrestic-rivos) [inline umode-api/src/lib.rs:177]: So all output is passed via an out pointer? Do you foresee having a scalar return value being useful for anything (e.g. to return the number of bytes written to the out pointer, or for ops that need to only return an integer)?

File: src/umode.rs [hunk 9]
```diff
@@ -231,81 +232,107 @@ global_asm!(
 pub enum Error {
     /// Received an unexpected trap while running Umode.
     UnexpectedTrap,
-    /// Task already active.
-    TaskBusy,
+    /// Umode called panic.
+    Panic,
```
Comment (abrestic-rivos) [inline src/umode.rs:236]: Should we poison the U-mode task after it panics? Or if it hits an unexpected trap?

File: libuser/src/lib.rs [hunk 10]
```diff
@@ -4,18 +4,53 @@
...
+        for c in s.chars() {
+            // Ignore errors for putchar.
+            let _ = hyp_putchar(c);
         }
+        Ok(())
     }
 }
+
+/// Global reference for `print` implementation.
+pub static mut UMODEWRITER: UserWriter = UserWriter {};
```
Comment (abrestic-rivos) [inline libuser/src/lib.rs:34]: nit: `UMODE_WRITER` or `USER_WRITER`, etc.

**Review by abrestic-rivos — COMMENTED** [review]
A couple minor comments, but overall LGTM -- thanks!

File: Cargo.toml (hunk 5 above)
Comment (glg-rv) [reply Cargo.toml:34]: :-) And rightly so. Will fix.

File: umode-api/src/lib.rs (hunk 6 above)
Comment (glg-rv) [reply umode-api/src/lib.rs:43]: Agreed.

File: umode-api/src/lib.rs (hunk 8 above)
Comment (glg-rv) [reply umode-api/src/lib.rs:177]: Yes, I will need something like this, and that can be changed easily. I just wanted to keep this to a minimum at the moment. Getting the basic infrastructure right first.

File: umode-api/src/lib.rs (hunk 7 above)
Comment (glg-rv) [reply umode-api/src/lib.rs:10]: Will do. Truth to be told I don't know where all this documentation goes.

File: src/umode.rs (hunk 9 above)
Comment (glg-rv) [reply src/umode.rs:236]: This is where the outstanding patch of reset comes along. Reset will be called exactly when unexpected trap, panic (or even unknown hypcall) is detected as an error after `task.run()` in `send_req`.

My plan so far is the following:
1. `umode-api` and basic umode handling lifetime <- this patch
2. Expanding `umode-api` with ability to map pages in input/output when running `umode`
3. implement reset on panic/error.

After this, the feature should be completed. A baseline patch for 3 is already completed, so I could anticipate it.

File: libuser/src/lib.rs (hunk 10 above)
Comment (glg-rv) [reply libuser/src/lib.rs:34]: Ack.

Comment (glg-rv) [reply libuser/src/lib.rs]: Fixed. This was actually a spurious line, I didn't need this static variable anymore -- rust missed that being declared public in a library.

File: src/umode.rs (hunk 9 above)
Comment (glg-rv) [reply src/umode.rs:236]: No changes regarding this. (still agree)

File: umode-api/src/lib.rs (hunk 8 above)
Comment (glg-rv) [reply umode-api/src/lib.rs]: No changes.

File: umode-api/src/lib.rs (hunk 7 above)
Comment (glg-rv) [reply umode-api/src/lib.rs]: Done.

File: umode-api/src/lib.rs (hunk 6 above)
Comment (glg-rv) [reply umode-api/src/lib.rs]: Done.

File: Cargo.toml (hunk 5 above)
Comment (glg-rv) [reply Cargo.toml]: Changed `umode` and `umode-api` to their hyphened counterpart.

File: src/umode.rs (hunk 9 above)
Comment (abrestic-rivos) [reply src/umode.rs:236]: Maybe drop a TODO to indicate that this is forthcoming?

**Review by abrestic-rivos — APPROVED** [review]

File: src/umode.rs (hunk 9 above)
Comment (glg-rv) [reply src/umode.rs:236]: Done. Things are slightly more complex that the comment but that's the essence.

**Comment by glg-rv** [comment]
Forgot to push the TODO change. All done now.

**Pull request merged**